Add should_traverse_fn to torch.fx.node.map_aggregate

pbelevich opened the pull request:
Adds an optional callback that checks if map_aggregate should continue recursive traversal. The main motivation is to not traverse torch.Size which is tuple

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81510

pbelevich committed:
Add should_traverse_fn to torch.fx.node.map_aggregate

[ghstack-poisoned]

pbelevich committed:
Update on "Add should_traverse_fn to torch.fx.node.map_aggregate"




[ghstack-poisoned]

jamesr66a commented on torch/fx/node.py:
```diff
@@ -601,19 +601,21 @@ def map_arg(a: Argument, fn: Callable[[Node], Argument]) -> Argument:
+    elif isinstance(a, dict) and (should_traverse_fn is None or should_traverse_fn(a)):
+        return immutable_dict((k, map_aggregate(v, fn, should_traverse_fn)) for k, v in a.items())
+    elif isinstance(a, slice) and (should_traverse_fn is None or should_traverse_fn(a)):
+        return slice(map_aggregate(a.start, fn, should_traverse_fn), map_aggregate(a.stop, fn, should_traverse_fn), map_aggregate(a.step, fn, should_traverse_fn))
```
How about:

```
    if should_traverse_fn and not should_traverse_fn(a):
      return fn(a)

    if isinstance(a, tuple):
        t = tuple(map_aggregate(elem, fn) for elem in a)
        # Support NamedTuple (if it has `_fields`) by repacking into original type.
        return t if not hasattr(a, '_fields') else type(a)(*t)
    elif isinstance(a, list):
        return immutable_list(map_aggregate(elem, fn) for elem in a)
    elif isinstance(a, dict):
        return immutable_dict((k, map_aggregate(v, fn)) for k, v in a.items())
    elif isinstance(a, slice):
        return slice(map_aggregate(a.start, fn), map_aggregate(a.stop, fn), map_aggregate(a.step, fn))

```

pbelevich replied:
addressed, thanks!

pbelevich committed:
Update on "Add should_traverse_fn to torch.fx.node.map_aggregate"




[ghstack-poisoned]

pbelevich committed:
Update on "Add should_traverse_fn to torch.fx.node.map_aggregate"

Adds an optional callback that checks if map_aggregate should continue recursive traversal. The main motivation is to not traverse torch.Size which is tuple




[ghstack-poisoned]

pbelevich committed:
Update on "Add should_traverse_fn to torch.fx.node.map_aggregate"

Adds an optional callback that checks if map_aggregate should continue recursive traversal. The main motivation is to not traverse torch.Size which is tuple




[ghstack-poisoned]

pbelevich commented:
@pytorchbot merge -g

pbelevich committed:
Update on "Add should_traverse_fn to torch.fx.node.map_aggregate"

Adds an optional callback that checks if map_aggregate should continue recursive traversal. The main motivation is to not traverse torch.Size which is tuple




[ghstack-poisoned]

pbelevich commented:
@pytorchbot merge -g
